src/scanner.c

```diff
@@ -0,0 +1,655 @@
+    lexer->advance(lexer, false);
+    if (lexer->lookahead == '>') {
+        lexer->advance(lexer, false);
+        if (scanner->tags.len > 0) {
+            VEC_POP(scanner->tags);
+            lexer->result_symbol = SELF_CLOSING_TAG_DELIMITER;
+        }
+        return true;
+    }
+    return false;
+}
+
+bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
+    while (iswspace(lexer->lookahead)) {
+        lexer->advance(lexer, true);
+    }
+
+    if (valid_symbols[RAW_TEXT] && !valid_symbols[START_TAG_NAME] &&
+        !valid_symbols[END_TAG_NAME]) {
+        return scan_raw_text(scanner, lexer);
+    }
+
+    switch (lexer->lookahead) {
+        case '<':
+            lexer->mark_end(lexer);
+            lexer->advance(lexer, false);
+
+            if (lexer->lookahead == '!') {
+                lexer->advance(lexer, false);
+                return scan_comment(lexer);
+            }
+
+            if (valid_symbols[IMPLICIT_END_TAG]) {
+                return scan_implicit_end_tag(scanner, lexer);
+            }
+            break;
+
+        case '\0':
+            if (valid_symbols[IMPLICIT_END_TAG]) {
+                return scan_implicit_end_tag(scanner, lexer);
+            }
+            break;
+
+        case '/':
+            if (valid_symbols[SELF_CLOSING_TAG_DELIMITER]) {
+                return scan_self_closing_tag_delimiter(scanner, lexer);
+            }
+            break;
+
+        case '{':
+            if (valid_symbols[INTERPOLATION_START]) {
+                lexer->advance(lexer, false);
+                /* tags.push_back(Tag(INTERPOLATION, string())); */
+                Tag tag = (Tag){INTERPOLATION, {0}};
+                VEC_PUSH(scanner->tags, tag);
+                lexer->result_symbol = INTERPOLATION_START;
+                return true;
+            }
+            break;
+
+        case '-':
+            if (valid_symbols[FRONTMATTER_START]) {
+                lexer->mark_end(lexer);
+                lexer->advance(lexer, false);
+                if (lexer->lookahead == '-') {
+                    lexer->advance(lexer, false);
+                    if (lexer->lookahead == '-') {
+                        lexer->advance(lexer, false);
+                        lexer->mark_end(lexer);
+                        lexer->result_symbol = FRONTMATTER_START;
+                        return true;
+                    }
+                }
+            }
+
+        default:
+            if ((valid_symbols[START_TAG_NAME] ||
+                 valid_symbols[END_TAG_NAME]) &&
+                !valid_symbols[RAW_TEXT]) {
+                return valid_symbols[START_TAG_NAME]
+                           ? scan_start_tag_name(scanner, lexer)
+                           : scan_end_tag_name(scanner, lexer);
+            }
+    }
+
+    return false;
+}
+
+void *tree_sitter_astro_external_scanner_create() {
+    Scanner *scanner = (Scanner *)calloc(1, sizeof(Scanner));
```

Comment:
`calloc()` _works_ here, but ideally you'd define a `VEC_INIT` macro and instead write:

```c
Scanner *scanner = malloc(sizeof(Scanner));
VEC_INIT(scanner->tags);
```

This way, if you ever change the format of Vec, you know that this code will keep working without having to change it. This can also be used for speed gains, e.g. `VEC_INIT` could allocate a starting buffer of size 16 to handle the most common case.